app/routes.py

```diff
@@ -1,2 +1,338 @@
+                    "description": task.description,
+                    "is_complete": False if task.completed_at == None else True
+                }
+            }
+        else:
+            # return jsonify(task), 200
+             return {
+                "task": {
+                    "id": task.task_id,
+                    "goal_id": task.goal_id,
+                    "title": task.title,
+                    "description": task.description,
+                    "is_complete": False if task.completed_at == None else True
+                }
+            }
+
+    elif request.method == "PUT":
+        if task is None:
+            return jsonify(None), 404
+        else:
+            form_data = request.get_json()
+
+            task.title = form_data["title"]
+            task.description = form_data["description"]
+
+            db.session.commit()
+
+        # return make_response(f"Book #{book.id} successfully updated")
+            return {
+                    "task": {
+                        "id": task.task_id,
+                        "title": task.title,
+                        "description": task.description,
+                        "is_complete": False if task.completed_at == None else True
+                    }
+                }
+        
+
+
+
+# def handle_book(book_id):
+#     for book in books:
+#         if book.id == book_id:
+#            return {
+#               "id": book.id,
+#               "title": book.title,
+#               "description": book.description
+#           }
+
+#  return {
+#         "id": book.id,
+#         "title": book.title,
+#         "description": book.description
+#     }
+
+   
+
+    if request.method == "DELETE":
+        if task is None:
+            return jsonify(None), 404
+        # request_body = request.get_json() request body contains information to help conduct the actual request
+        # ie post request requires data to determines how to actually create the task
+        db.session.delete(task)
+        db.session.commit()
+
+        return {"details": f'Task {task.task_id} "{task.title}" successfully deleted'}
+
+@tasks_bp.route("/<task_id>/mark_complete", methods= ["PATCH"])
+#whatever function underneath will run when this route is called
+#put updates the entire record and patch is used when you want to modify a particular property of a record
+
+def mark_complete(task_id):
+
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+    
+    title_of_task =  f"Someone just completed the task {task.title}" 
+    
+    requests.post("https://slack.com/api/chat.postMessage", data={"channel": "task-notifications", "channelid": "C02K8MA17KR", "text": title_of_task}, headers={"authorization": os.environ.get("SLACK_API_TOKEN")})
+    #we have to use the os.environ.get to get the value of the variable becacuse this is an environment variable. We know it's an environment variable because we created it in env file - 
```

Comment:
There are two parts to this function: (1) marking the task as complete in our database and (2) pinging the slackbot. To practice the single responsibility principle, we can move the logic for the slack bot into a helper function. This will also allow the slack bot to be used by other routes too (ex: confirming post, put, or get requests, etc.)